text/0000-forget-marker-trait.md

```diff
@@ -0,0 +1,803 @@
+
+```rust
+async fn something_with_clean_up(f: impl AsyncFnOnce(Foo)) {
+    // setup
+    f(Foo).await;
+    // cleanup
+}
+
+async fn main() {
+    let fut = something_with_clean_up(async |foo| {
+        foo.bar().await;
+    });
+    {
+        // Pin the future.
+        let pinned = Box::pin(fut);
+        // Poll the future once.
+        poll_fn(|cx| Poll::Ready(_ = pinned.poll(cx))).await;
+        forget(pinned); // or `_ = Box::leak(pinned);`
+    }
+    // rest of the code...
+}
+```
+
+The library is only taking control flow in between `await` points. Here, future is pinned and `Pin`'s [drop guarantee] is met (boxed future remains allocated for `'static`), but cleanup cannot run - `Drop` handler of `fut` is skipped. Thus, APIs that require any cleanup for safety can be expressed in `sync` Rust, but not in `async` Rust, making `async` less attractive, as the operating system APIs and C/C++ libraries *cannot* be used efficiently, ergonomically, and safely.
+
+[drop guarantee]: https://doc.rust-lang.org/std/pin/#drop-guarantee
+
+Another important observation that we can make is that `Pin`'s drop guarantee only applies to the memory of the `Future` itself. But if `Future` borrows a buffer, that buffer *can* be deallocated or re-used before the `drop` of the `Future` is called. See [#connection-to-pin](#connection-to-pin).
+
+## Examples of unsafe async APIs that can be allowed in sync Rust
+[example-safe-sync-unsafe-async]: #example-safe-sync-unsafe-async
+
+### Async spawn
+[example-async-spawn]: #example-async-spawn
+
+Example from the ecosystem: [spawn_unchecked][spawn_unchecked-example-doc]
+
+[spawn_unchecked-example-doc]: https://docs.rs/async-task/latest/async_task/fn.spawn_unchecked.html.
+
+With the `Forget` trait we can make that API safe:
+
+```rust
+struct TaskHandler<'a>(u64, PhantomNonForget, PhantomData<&'a ()>);
+// Or `struct TaskHandler<'a>(u64, PhantomNonForget<&'a ()>)`;
+
+impl Drop for TaskHandler<'_> {
+    fn drop(&mut self) {
+        if let Some(mut mutex) = GLOBAL.get(self.0) {
+            // We can block in async context as this mutex is held
+            // during the `poll` which should return in a timely manner.
+            let fut = mutex.lock();
+            // cancel the future and call its drop handler
+            drop(fut.take())
+        }
+    }
+}
+
+// Note that this is basically equivalent to async `scope`, as async `scope`
+// would be transformed into the `Future` struct, just like `TaskHandler`.
+fn spawn<'a>(fut: impl IntoFuture + 'a) -> TaskHandler<'a> {
+    GLOBAL.spawn(fut)
+}
+```
+
+### Async DMA
+[example-async-dma]: #example-async-dma
+
+DMA stands for Direct Memory Access and it’s a peripheral used for transferring data between two memory locations in parallel to the operation of the core processor. For the purposes of this example, it can be thought of as `memcpy` in parallel to any other code.
```

Comment:
DMA is used by various parts of the hardware, not just peripherals

```suggestion
DMA stands for Direct Memory Access, which is used to transfer data between two memory locations in parallel to the operation of the core processor. For the purposes of this example, it can be thought of as `memcpy` in parallel to any other code.
```